File: app/agent/funmax.py
from datetime import datetime
from typing import Any, Optional, Union

from pydantic import BaseModel, model_validator

from app.agent.react import ReActAgent
from app.context.browser import BrowserContextHelper
from app.context.toolcall import ToolCallContextHelper
from app.logger import logger
from app.prompt.funmax import NEXT_STEP_PROMPT, PLAN_PROMPT, SYSTEM_PROMPT
from app.schema import Message
from app.tool import Terminate, ToolCollection
from app.tool.base import BaseTool
from app.tool.bash import Bash
from app.tool.browser_use_tool import BrowserUseTool
from app.tool.create_chat_completion import CreateChatCompletion
from app.tool.deep_research import DeepResearch
from app.tool.file_operators import FileOperator
from app.tool.planning import PlanningTool
from app.tool.str_replace_editor import StrReplaceEditor
from app.tool.web_search import WebSearch
from app.utils.agent_event import BaseAgentEvents
from app.utils.prompt_template import template_manager


SYSTEM_TOOLS: list[BaseTool] = [
    Bash(),
    WebSearch(),
    DeepResearch(),
    BrowserUseTool(),
    FileOperator(),
    StrReplaceEditor(),
    PlanningTool(),
    CreateChatCompletion(),
]

SYSTEM_TOOLS_MAP = {tool.name: tool.__class__ for tool in SYSTEM_TOOLS}


class McpToolConfig(BaseModel):
    id: str
    name: str
    # for stdio
    command: str
    args: list[str]
    env: dict[str, str]
    # for sse
    url: str
    headers: dict[str, Any]


class FunMax(ReActAgent):
    """A versatile general-purpose agent."""

    name: str = "FunMax"
    description: str = (
        "A versatile agent that can solve various tasks using multiple tools"
    )

    language: str
    tools: list[Union[McpToolConfig, str]]
    task_request: str
    history: list[dict[str, Any]]
    custom_prompt_templates: dict[str, str] = {}

    # Override parent class fields to make them non-optional
    system_prompt: str = ""
    next_step_prompt: str = ""

    _tool_call_context_helper: Optional[ToolCallContextHelper] = None
    _browser_context_helper: Optional[BrowserContextHelper] = None

    @property
    def tool_call_context_helper(self) -> ToolCallContextHelper:
        if self._tool_call_context_helper is None:
            raise RuntimeError("tool_call_context_helper is not initialized")
        return self._tool_call_context_helper

    @property
    def browser_context_helper(self) -> BrowserContextHelper:
        if self._browser_context_helper is None:
            raise RuntimeError("browser_context_helper is not initialized")
        return self._browser_context_helper

    @model_validator(mode="after")
    def initialize_helper(self) -> "FunMax":
        organization_id, task_id = self.task_id.split("/")
        self.task_dir = f"/workspace/{organization_id}/{task_id}"

        self.system_prompt_template = self.custom_prompt_templates.get(
            "system_prompt", SYSTEM_PROMPT
        )
        self.next_step_prompt_template = self.custom_prompt_templates.get(
            "next_step_prompt", NEXT_STEP_PROMPT
        )
        self.plan_prompt_template = self.custom_prompt_templates.get(
            "plan_prompt", PLAN_PROMPT
        )

        self.system_prompt = template_manager.render_template_safe(
            self.system_prompt_template,
            task_id=task_id,
            name=self.name,
            language=self.language or "English",
            max_steps=self.max_steps,
            current_time=datetime.now().strftime("%Y-%m-%d %H:%M:%S UTC"),
        )
        self.next_step_prompt = template_manager.render_template_safe(
            self.next_step_prompt_template,
            max_steps=self.max_steps,
            current_step=self.current_step,
            remaining_steps=self.max_steps - self.current_step,
        )
        return self

    async def prepare(self) -> None:
        """Prepare the agent for execution."""
        await super().prepare()

        await self.update_memory(
            role="system", content=self.system_prompt, base64_image=None
        )

        if self.history:
            for message in self.history:
                await self.update_memory(
                    role=message["role"], content=message["message"]
                )

        self._browser_context_helper = BrowserContextHelper(self)
        self._tool_call_context_helper = ToolCallContextHelper(self)
        self.tool_call_context_helper.available_tools = ToolCollection(Terminate())

        if self.tools:
            for tool in self.tools:
                if isinstance(tool, str) and tool in SYSTEM_TOOLS_MAP:
                    inst = SYSTEM_TOOLS_MAP[tool]()
                    await self.tool_call_context_helper.add_tool(inst)
                    if hasattr(inst, "llm"):
                        inst.llm = self.llm
                    if hasattr(inst, "sandbox"):
                        inst.sandbox = self.sandbox
                elif isinstance(tool, McpToolConfig):
                    await self.tool_call_context_helper.add_mcp(
                        {
                            "client_id": tool.id,
                            "url": tool.url,
                            "command": tool.command,
                            "args": tool.args,
                            "env": tool.env,
                            "headers": tool.headers,
                        }
                    )
        summary = await self.llm.ask(
            [Message.user_message(self.task_request)],
            [
                Message.system_message(
                    "Summarize the requirements or tasks provided by the user, Ensure that the core of the task can be reflected, answer in the user's language within 15 characters"
                )
            ],
        )
        self.emit(BaseAgentEvents.LIFECYCLE_SUMMARY, {"summary": summary})
        print(summary)
        print("--------------------------------")
        print(
            f"prepare success, available tools: {', '.join([tool.name for tool in self.tool_call_context_helper.available_tools.tools])}"
        )

    async def plan(self) -> str:
        """Create an initial plan based on the user request."""
        # Create planning message
        self.emit(BaseAgentEvents.LIFECYCLE_PLAN_START, {})

        plan_prompt = template_manager.render_template_safe(
            self.plan_prompt_template,
            language=self.language or "English",
            max_steps=self.max_steps,
            available_tools="\n".join(
                [
                    f"- {tool.name}: {tool.description}"
                    for tool in self.tool_call_context_helper.available_tools
                ]
            ),
        )
        planning_message = await self.llm.ask(
            [
                Message.system_message(plan_prompt),
                Message.user_message(self.task_request),
            ],
            system_msgs=[Message.system_message(self.system_prompt)],
        )

        # Add the planning message to memory
        await self.update_memory("user", planning_message)
        self.emit(BaseAgentEvents.LIFECYCLE_PLAN_COMPLETE, {"plan": planning_message})
        return planning_message

    async def think(self) -> bool:
        """Process current state and decide next actions with appropriate context."""
        # Update next_step_prompt with current step information
        original_prompt = self.next_step_prompt
        self.next_step_prompt = template_manager.render_template_safe(
            self.next_step_prompt_template,
            max_steps=self.max_steps,
            current_step=self.current_step,
            remaining_steps=self.max_steps - self.current_step,
        )

        browser_in_use = self._check_browser_in_use_recently()

        if browser_in_use:
            self.next_step_prompt = (
                await self.browser_context_helper.format_next_step_prompt()
            )

        result = await self.tool_call_context_helper.ask_tool()

        # Restore original prompt
        self.next_step_prompt = original_prompt

        return result

    async def act(self) -> str:
        """Execute decided actions"""
        results = await self.tool_call_context_helper.execute_tool()
        return "\n\n".join(results)

    def _check_browser_in_use_recently(self) -> bool:
        """Check if the browser is in use by looking at the last 3 messages."""
        recent_messages = self.memory.messages[-3:] if self.memory.messages else []
        browser_in_use = any(
            tc.function.name == BrowserUseTool().name
            for msg in recent_messages
            if msg.tool_calls
            for tc in msg.tool_calls
        )
        return browser_in_use

    async def cleanup(self):
        """Clean up agent resources."""
        logger.info(f"🧹 Cleaning up resources for agent '{self.name}'...")
        if self.browser_context_helper:
            await self.browser_context_helper.cleanup_browser()
        if self.tool_call_context_helper:
            await self.tool_call_context_helper.cleanup_tools()
        await super().cleanup()
        logger.info(f"✨ Cleanup complete for agent '{self.name}'.")

File: packages/agent/src/agent/funmax.ts
import { UnifiedChat } from '@repo/llm/chat';
import NEXT_STEP_PROMPT from '../prompts/next';
import SYSTEM_PROMPT from '../prompts/system';
import { ToolCallContextHelper, ToolExecutionProgress, ToolSelectionProgress } from '../tools/toolcall';
import type { AddMcpConfig } from '../tools/types';
import { createMessage } from '../utils/message';
import { renderTemplate } from '../utils/template';
import { ReActAgent, ReActAgentConfig } from './react';
import { BaseAgentEvents } from './base';

export interface FunMaxConfig extends ReActAgentConfig {
  task_request: string;
  language?: string;
  tools?: AddMcpConfig[];
  history?: UnifiedChat.Message[];
  systemPromptTemplate?: string;
}

/**
 * FunMax - 一个多功能通用代理
 * 可以使用多种工具解决各种任务的通用代理
 */
export class FunMax extends ReActAgent {
  public readonly name: string = 'FunMax';
  public readonly description: string = 'A versatile agent that can solve various tasks using multiple tools';

  // 配置属性
  public readonly language: string;
  public readonly tools: AddMcpConfig[];
  public readonly task_request: string;
  public readonly history: UnifiedChat.Message[];
  public readonly system_prompt_template: string;

  // 上下文助手（暂时设为可选，实际使用时需要初始化）
  private _tool_call_context_helper?: ToolCallContextHelper;

  constructor(config: FunMaxConfig) {
    super(config);

    this.language = config.language || 'English';
    this.tools = config.tools || [];
    this.task_request = config.task_request;
    this.history = config.history || [];
    this.system_prompt_template = config.systemPromptTemplate || SYSTEM_PROMPT;
  }

  /**
   * 准备代理执行
   */
  public async *prepare(): AsyncGenerator<{ phase: string; data: any }, void, unknown> {
    yield { phase: BaseAgentEvents.LIFECYCLE_PREPARE_START, data: {} };
    yield* super.prepare();

    // 添加系统提示词到内存
    const system_prompt = renderTemplate(this.system_prompt_template, {
      language: this.language || 'English',
      max_steps: this.max_steps,
      current_time: new Date().toISOString(),
      workspace_dir: '/heyfun/workspace',
    });
    await this.updateMemory(createMessage.system(system_prompt));

    // 添加历史消息到内存
    if (this.history && this.history.length > 0) {
      for (const message of this.history) {
        await this.updateMemory(message);
      }
    }

    // 初始化上下文助手
    this._tool_call_context_helper = new ToolCallContextHelper(this);

    // 初始化工具集合
    yield { phase: BaseAgentEvents.LIFECYCLE_PREPARE_PROGRESS, data: { progress: 'Initializing tools...' } };
    await this.initializeTools();
    yield { phase: BaseAgentEvents.LIFECYCLE_PREPARE_PROGRESS, data: { progress: 'Tools initialized' } };
    yield { phase: BaseAgentEvents.LIFECYCLE_PREPARE_COMPLETE, data: {} };
  }

  /**
   * 初始化工具
   */
  private async initializeTools(): Promise<void> {
    if (!this._tool_call_context_helper) return;
    await this._tool_call_context_helper.initiate();

    const map = new Map<string, AddMcpConfig>();
    for (const tool of this.tools) {
      map.set(tool.id, tool);
    }

    // 添加系统工具和MCP工具;
    for (const tool of this.tools) {
      await this._tool_call_context_helper.addMcp(tool);
    }
  }

  /**
   * 生成任务摘要
   */
  async generateTaskSummary(): Promise<string> {
    try {
      const response = await this.llm.chat({
        messages: [
          createMessage.system(
            "Generate a concise title that captures the essence of the user's request or task. Do not answer the question or provide solutions - only create a descriptive title. Use the user's language and keep it within 15 characters.",
          ),
          createMessage.user(this.task_request),
        ],
      });

      const content = response.choices[0]?.message?.content;
      return typeof content === 'string' ? content : 'Task summary not available';
    } catch (error: any) {
      console.error('Error generating task summary:', error?.message || error);
      return 'Unable to generate task summary';
    }
  }

  /**
   * 思考阶段 - 处理当前状态并决定下一步行动
   */
  public async *think(): AsyncGenerator<ToolSelectionProgress, { prompt: string; result: boolean }, unknown> {
    // 更新下一步提示词
    const next_step_prompt = renderTemplate(NEXT_STEP_PROMPT, {
      max_steps: this.max_steps,
      current_step: this.current_step,
      remaining_steps: this.max_steps - this.current_step,
      language: this.language || 'English',
    });

    let result = false;
    if (this._tool_call_context_helper) {
      const askToolStream = this._tool_call_context_helper.askToolStream(next_step_prompt);
      let iteratorResult;

      while (!(iteratorResult = await askToolStream.next()).done) {
        const toolSelectionProgress = iteratorResult.value;
        yield toolSelectionProgress;
      }

      result = iteratorResult.value;
    }

    return { prompt: next_step_prompt, result };
  }

  /**
   * 行动阶段 - 执行决定的行动（流式版本）
   */
  public async *act(): AsyncGenerator<ToolExecutionProgress, string, unknown> {
    if (!this._tool_call_context_helper) {
      return 'No tool context helper available';
    }

    const toolStream = this._tool_call_context_helper.executeToolStream();
    let iteratorResult;

    while (!(iteratorResult = await toolStream.next()).done) {
      const progress = iteratorResult.value;
      yield progress;
    }

    const results = iteratorResult.value;
    return results.join('\n\n');
  }

  /**
   * 清理资源
   */
  protected async cleanup(): Promise<void> {
    if (this._tool_call_context_helper) {
      await this._tool_call_context_helper.cleanup();
    }
    await super.cleanup();
  }
}
